enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -4715,7 +4721,279 @@ class AdjointGenerator
     }
   }
 
-  bool handleBLAS(llvm::CallInst &call, Function *called, StringRef funcName,
+  std::string getBLASName(StringRef funcName, const std::string &deriv) {
+    return funcName.str().substr(0, 7) + deriv;
+  }
+
+  Type *getBLASInnerType(StringRef funcName, LLVMContext &context) {
+    char scalarTypeName = funcName.str()[6];
+    Type *scalarType = nullptr;
+    switch (scalarTypeName) {
+    case 's':
+      scalarType = Type::getFloatTy(context);
+      break;
+    case 'd':
+      scalarType = Type::getDoubleTy(context);
+      break;
+    default:
+      assert(false && "Complex numbers are not handled");
+    }
+    return scalarType;
+  }
+
+  enum CBLAS_ORDER { CblasRowMajor = 101, CblasColMajor = 102 };
+
+  enum CBLAS_TRANSPOSE {
+    CblasNoTrans = 111,
+    CblasTrans = 112,
+    CblasConjTrans = 113,
+    AtlasConj = 114
+  };
+
+  enum BLASType { Void, Integer, Real, RealPointer };
+  struct BLASArgument {
+    std::string name;
+    BLASType type;
+  };
+  enum GradientOpType { BLASCall, HelperCall, AddToDiffe };
+  struct GradientOp {
+    GradientOpType gradientOpType;
+    std::string callee;
+    BLASType returnType;
+    std::vector<std::string> args;
+  };
+  struct ChainRule {
+    std::vector<GradientOp> gradientOps;
+  };
+  struct BLASAdjointGenerator {
+    std::string name;
+    BLASType returnType;
```

Comment:
There's so much going on here that I'm inclined to put this in a separate BLAS file